dojo-dungeon/src/dojo/rooms/rooms/riddle/Fragen_Pattern.java

```diff
@@ -103,25 +106,37 @@ private BiConsumer<Task, Set<TaskContent>> showAnswersOnHud() {
       if (answer.matches(EXPECTED_PATTERNS[currentPatternIndex])) {
         OkDialog.showOkDialog("Ihre Antwort ist korrekt!", "Antwort", () -> {});
         correctAnswerCount++;
-        if (correctAnswerCount >= 2) {
+        if (correctAnswerCount == NUMBER_OF_CORRECT_ANSWERS_NEEDED) {
+          OkDialog.showOkDialog(
+              "Die Tür wird geöffnet, aber Sie können auch noch weiter Fragen beantworten!",
+              "Antwort",
+              () -> {});
           openDoors();
         }
-        currentPatternIndex++;
-        if (currentPatternIndex < EXPECTED_PATTERNS.length) {
-          setNextTask();
-        }
+        setNextTask();
       } else {
-        OkDialog.showOkDialog("Ihre Antwort ist nicht korrekt!", "Ok", () -> {});
+        decreaseHerosHealthAtWrongTry();
       }
     };
   }
 
   private Quiz newFreeText() {
+    nextPattern();
     String questionText =
         "Welches Design-Pattern wird in dem UML-Klassendiagramm unter \""
             + FILE_NAME_PREFIX
             + (currentPatternIndex + 1)
             + ".png\" dargestellt? Es reicht das Wort ohne den Zusatz Pattern!";
     return new FreeText(questionText);
   }
+
+  private void nextPattern() {
+    if (patternIndices == null || patternIndices.isEmpty()) {
+      List<Integer> tempList =
+          new ArrayList<>(IntStream.range(0, EXPECTED_PATTERNS.length).boxed().toList());
+      Collections.shuffle(tempList);
+      patternIndices = new ArrayDeque<>(tempList);
+    }
+    currentPatternIndex = patternIndices.remove();
```

Comment:
Das Entfernen von Elementen aus einer Datenstruktur ist immer teuer. Warum speicherst Du nicht den Index in diese Datenstruktur und erhöhst den beim nächsten Mal (+1, % length)?